homeassistant/components/file_upload/__init__.py

```diff
@@ -0,0 +1,176 @@
+class FileUploadData:
+    """File upload data."""
+
+    temp_dir: Path
+    files: dict[str, str]
+
+    @classmethod
+    async def create(cls, hass: HomeAssistant) -> FileUploadData:
+        """Initialize the file upload data."""
+
+        def _create_temp_dir() -> Path:
+            """Create temporary directory."""
+            temp_dir = Path(tempfile.gettempdir()) / TEMP_DIR_NAME
+
+            # If it exists, it's an old one and Home Assistant didn't shut down correctly.
+            if temp_dir.exists():
+                shutil.rmtree(temp_dir)
+
+            temp_dir.mkdir()
+            return temp_dir
+
+        temp_dir = Path(await hass.async_add_executor_job(_create_temp_dir))
+
+        def cleanup_unused_files(ev: Event) -> None:
+            """Clean up unused files."""
+            shutil.rmtree(temp_dir)
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cleanup_unused_files)
+
+        return cls(temp_dir, {})
+
+    def has_file(self, file_id: str) -> bool:
+        """Return if file exists."""
+        return file_id in self.files
+
+    def file_dir(self, file_id: str) -> Path:
+        """Return the file directory."""
+        return self.temp_dir / file_id
+
+    def file_path(self, file_id: str) -> Path:
+        """Return the file path."""
+        return self.file_dir(file_id) / self.files[file_id]
+
+
+class FileUploadView(HomeAssistantView):
+    """HTTP View to upload files."""
+
+    url = "/api/file_upload"
+    name = "api:file_upload"
+
+    _upload_lock: asyncio.Lock | None = None
+
+    @callback
+    def _get_upload_lock(self) -> asyncio.Lock:
+        """Get upload lock."""
+        if self._upload_lock is None:
+            self._upload_lock = asyncio.Lock()
+
+        return self._upload_lock
+
+    async def post(self, request: web.Request) -> web.Response:
+        """Upload a file."""
+        async with self._get_upload_lock():
+            return await self._upload_file(request)
+
+    async def _upload_file(self, request: web.Request) -> web.Response:
+        """Handle uploaded file."""
+        # Increase max payload
+        request._client_max_size = MAX_SIZE  # pylint: disable=protected-access
+
+        data = await request.post()
+        file_field = data.get("file")
+
+        if not isinstance(file_field, web.FileField):
+            raise vol.Invalid("Expected a file")
+
+        hass: HomeAssistant = request.app["hass"]
+        file_id = ulid_hex()
+
+        if DOMAIN not in hass.data:
+            hass.data[DOMAIN] = await FileUploadData.create(hass)
+
+        file_upload_data: FileUploadData = hass.data[DOMAIN]
+        file_dir = file_upload_data.file_dir(file_id)
+
+        def _sync_work() -> None:
+            file_dir.mkdir()
+
+            # MyPy forgets about the isinstance check because we're in a function scope
+            assert isinstance(file_field, web.FileField)
+
+            with (file_dir / file_field.filename).open("wb") as target_fileobj:
```

Comment:
Passing `file_field.filename == "../../../../../../any/path"` via a hand-crafted multipart upload will break out of the temp folder and allow for any user authorized to upload files to overwrite any file accessible by the HA process.

Is there a need to utilize the user-provided filename at all? Can we instead write file contents to `home-assistant-file_upload/{file_id}`, with no intermediate directories?

Reply:
Added a check to make sure `file_field.filename` is just a filename.